README.md
# subscribe_notify_test

It's test to JsSIP extension: SUBSCRIBE/NOTIFY dialogs.
Please see: [https://github.com/versatica/JsSIP/issues/708](https://github.com/versatica/JsSIP/issues/708)

Note:
It's JsSIP subscriber/notifier API test, not presence client.
To use the code as presence client, you should modify the subscriber constructor argumentsand create SUBSCRIBE with appropriate for your server :

- eventName ( probably 'presence' ?)
- accept - list of content-type that will be used in receiving NOTIFYs. 
- contentType - content-type of your SUBSCRIBE
- set some body (if need) to subscriber.subscribe(). (in my test I send without body)

To adopt the example for your case, you should take a SIP trace for some client working with your server,
and see how to build SUBSCRIBE request.
Modify the test, and check that it generate the SUBSCRIBE as you need.
To check generated SIP SUBSCRIBE message:
please open browser dev tool (Ctrl/I) and select 'console' tab. The send/receive SIP messages will be shown in the log.



CHANGELOG.md
Added 'pending' state

Checked event name with id (e.g. event 'test' can be replaced to 'test;id=123')

Rewriting with Subsciber/Notifier included to JsSIP code.


test.js

/*
  SUBSCRIBE/NOTIFY JsSIP test
 */
let server;
let account;
let jssipUA;
let subscriber = null;
let notifier = null;

// Run when document is ready
function main() {
    server = guiLoadServerConfig({ domain: '', addresses: '' });
    account = guiLoadAccount({ user: '', password: '', displayName: '', authUser: '' });
    init();
    guiShowPanel('main_panel');
}

function init() {
    document.getElementById('setting_btn').onclick = () => { guiShowPanel('setting_panel'); }
    document.getElementById('subscribe_test_btn').onclick = guiSubscribeTest;

    //----------- set server and account fields in HTML ------------
    document.querySelector('#setting [name=sip_domain]').value = server.domain;
    document.querySelector('#setting [name=sip_addresses]').value = JSON.stringify(server.addresses);
    document.querySelector('#setting [name=user]').value = account.user;
    document.querySelector('#setting [name=password]').value = account.password;
    document.querySelector('#setting [name=display_name]').value = account.displayName;
    document.querySelector('#setting [name=auth_user]').value = account.authUser;

    document.getElementById('login_btn').onclick = function () {
        let user = document.querySelector('#setting [name=user]').value || '';
        let authUser = document.querySelector('#setting [name=auth_user]').value || '';
        let password = document.querySelector('#setting [name=password]').value || '';
        let displayName = document.querySelector('#setting [name=display_name]').value || '';

        // trim spaces
        user = user.trim();
        authUser = authUser.trim();
        password = password.trim();
        displayName = displayName.trim();

        let account = {
            user: user,
            authUser: authUser,
            password: password,
            displayName: displayName
        };
        guiStoreAccount(account);

        let domain = document.querySelector('#setting [name=sip_domain]').value;
        let addresses = document.querySelector('#setting [name=sip_addresses]').value;

        // trim spaces
        domain = domain.trim();
        addresses = addresses.trim();

        let conf = {
            domain: domain,
            addresses: JSON.parse(addresses),
        };
        guiStoreServerConfig(conf);
        location.reload();
    }

    document.getElementById('subscribe_test_btn').onclick = guiSubscribeTest;
    document.getElementById('subscribe_return_btn').onclick = function () { guiInfo(''); guiShowPanel('main_panel'); }
    document.getElementById('send_init_subscribe_btn').onclick = guiSendInitSubscribe;
    document.getElementById('send_next_subscribe_btn').onclick = guiSendNextSubscribe;
    document.getElementById('send_unsubscribe_btn').onclick = guiSendUnsubscribe;
    document.getElementById('send_notify_btn').onclick = guiSendNotify;
    document.getElementById('send_final_notify_btn').onclick = guiSendFinalNotify;

    if (server.domain && server.addresses && account.user && account.password) {
        try {
            stackInit();
        } catch (e) {
            guiError(e);
            guiShowPanel('setting_panel');
            return;
        }
    } else {
        guiInfo('Please fill server & account');
        guiShowPanel('setting_panel');
    }

}

//----------------- Local storage load/store ----------------------
function guiLoadAccount(def) { return storageLoadConfig('testAccount', def); }
function guiStoreAccount(value) { storageSaveConfig('testAccount', value); }
function guiLoadServerConfig(def) { return storageLoadConfig('testServerConfig', def); }
function guiStoreServerConfig(value) { storageSaveConfig('testServerConfig', value); }
function storageLoadConfig(name, def = {}) {
    let str_value = localStorage.getItem(name);
    return str_value ? JSON.parse(str_value) : def;
}
function storageSaveConfig(name, value) {
    localStorage.setItem(name, JSON.stringify(value));
}

//------------- Set status line --------------------
function guiError(text) { guiStatus(text, 'Pink'); }
function guiWarning(text) { guiStatus(text, 'Gold'); }
function guiInfo(text) { guiStatus(text, 'Aquamarine'); }
function guiStatus(text, color) {
    let line = document.getElementById('status_line');
    line.setAttribute('style', `background-color: ${color}`);
    line.innerHTML = text;
}

//--------------- Show or hide element -------
function guiShow(id) {
    document.getElementById(id).style.display = 'block';
}
function guiHide(id) {
    document.getElementById(id).style.display = 'none';
}

//--------------- Show active panel and hide others  ----------------
function guiShowPanel(activePanel) {
    const panels = ['main_panel', 'setting_panel', 'subscribe_panel'
    ];
    for (let panel of panels) {
        if (panel === activePanel) {
            guiShow(panel);
        } else {
            guiHide(panel);
        }
    }
}

//----------------- JsSIP init ------------------------
function stackInit() {
    let sockets = [];
    for (let address of server.addresses) {
        if (address instanceof Array) { // 'address' or ['address', weight]
            sockets.push({ socket: new JsSIP.WebSocketInterface(address[0]), weight: address[1] });
        } else {
            sockets.push(new JsSIP.WebSocketInterface(address));
        }
    }

    let config = {
        sockets: sockets,
        uri: 'sip:' + account.user + '@' + server.domain,
        contact_uri: 'sip:' + account.user + '@' + JsSIP.Utils.createRandomToken(12) + '.invalid;transport=ws',
        authorization_user: account.authUser ? account.authUser : account.user,
        password: account.password,
        register: true,
        register_expires: 3600,
    };

    if (account.displayName && account.displayName.length > 0) {
        config.display_name = account.displayName;
    }

    jssipUA = new JsSIP.UA(config);
    jssipUA.on('connected', (e) => {
        guiInfo('connected');
    });

    jssipUA.on('disconnected', () => {
        guiInfo('disconnected');
        guiShowPanel('main_panel');
    });

    jssipUA.on('registered', (e) => {
        guiInfo('registered');
    });

    jssipUA.on('unregistered', (e) => {
        guiInfo('unregistered');
    });

    jssipUA.on('registrationFailed', (e) => {
        guiError('regisration failed');
    });

    jssipUA.on('newSubscribe', (e) => {
        let subs = e.request;
        let ev = subs.parseHeader('event');
        let accepts = subs.getHeaders('accept');
        console.log('incomingSubscribe', subs, ev.event, accepts);
        let code = incomingSubscribe(subs, ev.event, accepts);
        if (code > 0)
            subs.reply(code);
    });

    JsSIP.debug.enable('JsSIP:*');
    jssipUA.start();
}

//--------- incoming SUBSCRIBE. Accept or reject ?  ----------
function incomingSubscribe(subscribe, eventName, accepts) {
    // Check incoming SUBSCRIBE
    const ourEventName = 'test';
    const ourContentType = 'text/plain';

    // Check event type
    if (eventName !== ourEventName) {
        guiWarning('receive SUBSCRIBE: not supported our event');
        return 489; // send SIP response 489 Bad Event
    }
    // Check if accept header includes our content-type
    if (!accepts || !accepts.some(v => v.includes(ourContentType))) {
        guiWarning('receive SUBSCRIBE: accept header missed our content-type');
        return 406; // send SIP response 406 Not Acceptable.
    }
    try {
        createNotifier(subscribe);
        return 0; // Don't send SIP response. The created dialog send it.
    } catch (e) {
        guiWarning('Cannot notifier');
        console.log('Cannot notifier', e);
        return 400; // send SIP response 400 Bad Request.
    }
}

//--------------- Subscribe test panel -------------------------------
function guiSubscribeTest() {
    guiInfo('');
    guiShowPanel('subscribe_panel');
    guiShowButtons();
}

function guiShowButtons() {
    document.getElementById('send_init_subscribe_btn').disabled = !!subscriber;
    document.getElementById('send_next_subscribe_btn').disabled = !subscriber;
    document.getElementById('send_unsubscribe_btn').disabled = !subscriber;
    document.getElementById('send_notify_btn').disabled = !notifier;
    document.getElementById('send_final_notify_btn').disabled = !notifier;
}

//-------------------------------------------------------------------------
//--------------- subscriber (client subscribe dialog)  -------------------
//-------------------------------------------------------------------------
function guiSendInitSubscribe() {
    let sendToUser = document.querySelector('#send_subscribe_form [name=user]').value.trim();
    let eventName = document.querySelector('#send_subscribe_form [name=event_name]').value.trim();
    let expires = parseInt(document.querySelector('#send_subscribe_form [name=expires]').value.trim());

    if (sendToUser === '') {
        guiWarning('Missed user name');
        return;
    }

    let target = sendToUser;

    let params = null;
    /*
      params is optional.
      Used if domain or from-user is different from used in REGISTER/INVITE
  
      let params = {
          to_uri: new JsSIP.URI('sip', target, server.domain),
          to_display_name: null,
          from_uri: new JsSIP.URI('sip', account.user, server.domain),
          from_display_name: null,
      }
    */

    let credential = null;
    /* 
     credential is optional.
     Used if authentication is different from REGISTER/INVITE

     let credential = {
         authorization_user: phone.account.authUser ? phone.account.authUser : phone.account.user,
         password: phone.account.password
     };
    */

    try {
        subscriber = jssipUA.subscriber(target, {
            eventName,          // event name with optional ;id=xxx
            accept: 'text/json,text/plain', // We understand NOTIFY with the Content-Type
            expires: expires,               // Subscription expires. E.g. 3600
            contentType: 'text/plain',      // Content-Type of SUBSCRIBE requests.
            params: params,
            credential: credential,
        });
    } catch (e) {
        console.log('Error: cannot create subscriber', e);
        guiError('Cannot create subscriber');
    }

    /** 
     * 1st NOTIFY with Subscription-State: active
     */
    subscriber.on('active', () => {
        console.log('>> subscriber is active');
        guiInfo('subscriber: active');
    });

    /** 
        Incoming NOTIFY with body 
        If NOTIFY Subscription-State: terminated - the argument isTerminated = true 
    */
    subscriber.on('notify', (isFinal, notify, body, contentType) => { // with not empty body
        console.log(`>> receive ${isFinal ? 'final ' : ''}NOTIFY`, notify, body, contentType);
        guiInfo(`receive ${isFinal ? 'final ' : ''}notify`);
    });

    /**
     * subscriber dialog terminated. 
     * 
     * termination code converted to English text.
     * 
     * For terminationCode==RECEIVE_FINAL_NOTIFY 
     * can be set optional SubscriptionState header parameters:
     *  reason  (undefined or string)
     *  retryAfter (undefined or number)
     */
    subscriber.on('terminated', (terminationCode, reason, retryAfter) => {
        let terminationText = subscriberTerminationText(subscriber, terminationCode);
        console.log(`subscriber>>: terminated (${terminationText})${reason ? (' reason="' + reason + '"') : ''}${retryAfter !== undefined ? (' retry-after=' + retryAfter) : ''}`);
        guiWarning(`subscriber: terminated (${terminationText})${reason ? (' reason="' + reason + '"') : ''}`);
        subscriber = null;
        if( retryAfter !== undefined ){
            console.log(`You asked repeat subscription after ${retryAfter} seconds`);
        }
        guiShowButtons();
    });

    if (expires > 0) {
        // normal subscribe
        subscriber.subscribe();
    } else {
        // fetch SUBSCRIBE (with expires: 0), see RFC 6665 4.4.3
        subscriber.unsubscribe();
    }
    guiShowButtons();
}

// Convert termination code to English message.
function subscriberTerminationText(subscriber, terminationCode) {
    switch (terminationCode) {
        case subscriber.C.SUBSCRIBE_RESPONSE_TIMEOUT: return 'subscribe response timeout';
        case subscriber.C.SUBSCRIBE_TRANSPORT_ERROR: return 'subscribe transport error';
        case subscriber.C.SUBSCRIBE_NON_OK_RESPONSE: return 'subscribe non-OK response';
        case subscriber.C.SUBSCRIBE_FAILED_AUTHENTICATION: return 'subscribe failed authentication';
        case subscriber.C.UNSUBSCRIBE_TIMEOUT: return 'un-unsubscribe timeout';
        case subscriber.C.RECEIVE_FINAL_NOTIFY: return 'receive final notify';
        case subscriber.C.RECEIVE_BAD_NOTIFY: return 'receive bad notify';
        default: return 'unknown termination code: ' + terminationCode;
    }
}

// Send next SUBSCRIBE (after initial)
function guiSendNextSubscribe() {
    if (subscriber === null || subscriber.state === 'terminated') {
        guiWarning('No client subscribe dialog');
        return;
    }
    subscriber.subscribe('Next subscribe');
}

// Send unSubscribe (SUBSCRIBE with expires: 0)
function guiSendUnsubscribe() {
    if (subscriber === null || subscriber.state === 'terminated') {
        guiWarning('No client subscribe dialog');
        return;
    }
    subscriber.unsubscribe();
}

//-----------------------------------------------------------------------
//------------ notifier (server subscribe dialog)  ----------------------
//-----------------------------------------------------------------------
// In clients, it is used less often than subscriber
// In this test, it is used to debug the subscriber

function createNotifier(subscribe) {
    const ourContentType = 'text/plain';
    let pending = true; // notifier can be created in 'active' or 'pending' state

    // to test fetch 
    const isFetchSubscribe = subscribe.getHeader('expires') === '0';

    notifier = jssipUA.notifier({ subscribe, contentType: ourContentType, pending });

    // The event called for intitial and next subscribes.
    notifier.on('subscribe', (isUnsubscribe, subscribe, body, contentType) => {
        console.log(`notifier>> receive ${isUnsubscribe ? 'un-' : ''}SUBSCRIBE`, subscribe, body, contentType, isUnsubscribe);
        guiInfo('receive subscribe');
        if (isUnsubscribe) {
            notifier.sendFinalNotify(`Provide current system state (final notify)${isFetchSubscribe ? ' (fetch subscribe)' : ''}`);
        } else {
            if (notifier.state === 'pending') {
                notifier.sendNotify('State is pending. Do not provide system state');
            } else {
                notifier.sendNotify('Provide current system state');
            }
        }
    });

    notifier.on('terminated', (terminationCode, sendFinalNotify) => {
        let terminationText = notifierTerminationText(notifier, terminationCode);
        guiWarning(`notifier>> terminated (${terminationText})`);
        // sendFinalNotify=true for subscription timeout case
        // you have to send final NOTIFY in the case (with or without body)
        if (sendFinalNotify) {
            notifier.sendFinalNotify('Final notify. Provide current system state (if was)');
        }
        notifier = null;
        guiShowButtons();
    });

    notifier.start();
    guiShowButtons();
}

function notifierTerminationText(notifier, terminationCode) {
    switch (terminationCode) {
        case notifier.C.NOTIFY_RESPONSE_TIMEOUT: return 'notify response timeout';
        case notifier.C.NOTIFY_TRANSPORT_ERROR: return 'notify transport error';
        case notifier.C.NOTIFY_NON_OK_RESPONSE: return 'notify non-OK response';
        case notifier.C.NOTIFY_FAILED_AUTHENTICATION: return 'notify failed authentication';
        case notifier.C.SEND_FINAL_NOTIFY: return 'send final notify';
        case notifier.C.RECEIVE_UNSUBSCRIBE: return 'receive un-subscribe';
        case notifier.C.SUBSCRIPTION_EXPIRED: return 'subscription expired';
        default: return 'unknown termination code: ' + terminationCode;
    }
}

function guiSendNotify() {
    if (notifier === null || notifier.state === 'terminated') {
        guiWarning('No notifier');
        return;
    }
    // Switch state from pending to active.
    if (notifier.state === 'pending') {
        console.log('Switch state from "pending" to "active"');
        notifier.setActiveState();
    }
    notifier.sendNotify('Hi !');
}

function guiSendFinalNotify() {
    if (notifier === null || notifier.state === 'terminated') {
        guiWarning('No notifier');
        return;
    }
    // final notify
    // notifier.sendFinalNotify('final state');

    // final notify with reason and retry-after
    notifier.sendFinalNotify('final state', 'probation', 20);
}

utils.js
'use strict';

/*
 * Client SUBSCRIBE dialog
 */
class ClientSubscribeDialog {
    constructor({ jssipUA, log, target, eventName, accept, expires, contentType, allowEvents, params, headers, listeners, credential }) {
        if (!jssipUA)
            throw 'missed reference to JsSIP UA instance';
        this.jssipUA = jssipUA;
        this.log = log ? log : console.log;
        if (!target)
            throw 'missed target';
        this.target = target;
        if (!eventName)
            throw 'missed eventName';
        this.eventName = eventName;
        if (!accept)
            throw 'missed accept';
        this.accept = accept;
        if (!expires)
            expires = 900;
        this.expires = expires;
        this.allowEvents = allowEvents; // optional
        this.contentType = contentType; // optional
        if (!params)
            throw 'missed params';
        if (!params.from_uri)
            throw 'missed params.from_uri';
        if (!params.to_uri)
            throw 'missed params.to_uri';
        this.params = params;
        params.from_tag = JsSIP.Utils.newTag();
        params.to_tag = null;
        params.call_id = JsSIP.Utils.createRandomToken(20);
        params.cseq = Math.floor(Math.random() * 10000 + 1);
        this.contact = '<sip:' + params.from_uri.user + '@' + JsSIP.Utils.createRandomToken(12) + '.invalid;transport=ws>';
        //this.contact = '<sip:' + params.from_uri.user + '@' + params.from_uri.host + ';transport=ws>';
        if (!listeners)
            throw 'missed listeners';
        if (!listeners.active)
            throw 'missed listeners.active()';
        if (!listeners.notify)
            throw 'missed listeners.notify()';
        if (!listeners.terminated)
            throw 'missed listeners.terminated()';
        this.listeners = listeners;
        this.credential = credential; // optional
        this.state = 'init'; // init, notify_wait, pending, active, terminated
        this.id = null;      // dialog id
        this.expiresTimer = null;   // resend SUBSCRIBE before expiration
        this.expiresTS = null;      // expires timestamp.
        if (!headers)
            headers = [];
        this.headers = headers.concat([
            'Allow: SUBSCRIBE',
            'Event: ' + this.eventName,
            'Accept: ' + this.accept, // Content-type of NOTIFY body
            'Expires: ' + this.expires,
            'Contact: ' + this.contact,
        ]);
        if (this.allowEvents)
            this.headers.push('Allow-Events: ' + this.allowEvents);
        this.wasTerminated = false;
        this.route_set = null;
        this.data = {};
    }

    // SUBSCRIBE callbacks
    onAuthenticated() {
        this.params.cseq++;
    }
    onRequestTimeout() {
        this._dialogTerminated('subscribe response timeout');
    }
    onTransportError() {
        this._dialogTerminated('subscribe transport error');
    }
    onReceiveResponse(response) {
        if (response.status_code >= 200 && response.status_code < 300) {
            if (this.params.to_tag === null) {
                this.params.to_tag = response.to_tag;
                this.id = this.params.call_id + this.params.from_tag + this.params.to_tag;
                this.log('CSubs: added dialog id=' + this.id);
                this.jssipUA.newDialog(this);
                this.route_set = response.getHeaders('record-route').reverse();
                if (this.route_set.length > 0)
                    this.params.route_set = this.route_set;
            }
            let expires = this._getExpires(response);
            if (expires === -1) {
                this.log('CSubs: Error: OK without Expires header');
                return;
            }
            if (expires > 0) {
                this.expiresTS = new Date().getTime() + expires * 1000;
                let timeout = this._calculateTimeoutMs(expires);
                this._scheduleSubscribe(timeout);
            }
        } else if (response.status_code >= 300) {
            this._dialogTerminated('receive subscribe non-OK response');
        }
    }
    // end of SUBSCRIBE callbacks.

    // dialog callback
    receiveRequest(request) {
        if (request.method !== 'NOTIFY') {
            reply(405); // Method Not Allowed    
            return;
        }
        let subsState = request.parseHeader('subscription-state');
        if (!subsState) {
            this.log('CSubs: Error: NOTIFY without Subscription-State');
            reply(400); // Bad request
            return;
        }
        request.reply(200);

        // update state
        let newState = subsState.state.toLowerCase();
        let prevState = this.state;

        if (prevState !== 'terminated' && newState !== 'terminated') {
            this.state = newState;
            if (subsState.expires !== undefined) {
                let expires = subsState.expires;
                let notifyExpiresTS = new Date().getTime() + expires * 1000;
                const MaxTimeDeviation = 4000;
                // Shorter and the difference is too big
                if (this.expiresTS - notifyExpiresTS > MaxTimeDeviation) {
                    this.log('CSubs: update sending re-SUBSCRIBE time');
                    clearTimeout(this.expiresTimer);
                    this.expiresTS = notifyExpiresTS;
                    let timeout = this._calculateTimeoutMs(expires);
                    this._scheduleSubscribe(timeout);
                }
            }
        }
        // active callback (dialog state switched to active)
        if (prevState !== 'active' && newState === 'active') {
            this.log('CSubs>>> active: id=' + this.id);
            this.listeners.active(this);
        }

        // notify callback called only if NOTIFY has body
        let body = request.body;
        let isFinal = newState === 'terminated';
        if (body) {
            let ct = request.getHeader('content-type');
            this.log(`CSubs>>> ${isFinal ? 'final ' : ''} notify: id=` + this.id, body, ct);
            this.listeners.notify(this, isFinal, request, body, ct);
        }
        if (isFinal)
            this._dialogTerminated('receive final notify');
    }

    subscribe(body = null) {
        if (this.state === 'init')
            this.state = 'notify_wait';
        let headers = this.headers.slice();
        if (body) {
            if (!this.contentType) {
                this.log('CSubs: subscribe(): Error - contentType is not defined');
                throw 'CSubs.subscribe(): Missed contentType';
            }
            headers.push('Content-Type: ' + this.contentType);
        }
        this._send(body, headers);
    }

    unsubscribe() {
        this._dialogTerminated('send un-subscribe');
        let headers = [
            'Event: ' + this.eventName,
            'Expires: 0'
        ];
        this._send(null, headers);
    }

    _dialogTerminated(reason) {
        if (this.wasTerminated) // to prevent duplicate call
            return;
        this.wasTerminated = true;
        this.state = 'terminated';
        clearTimeout(this.expiresTimer);
        // remove dialog from dialogs table with some delay, to allow receive end NOTIFY
        setTimeout(() => {
            this.log('CSubs: removed dialog id=' + this.id);
            this.jssipUA.destroyDialog(this);
        }, 32000);
        this.log(`CSubs>>> terminated: "${reason}" id=${this.id}`);
        this.listeners.terminated(this, reason);
    }
    _send(body, headers) {
        this.params.cseq++;
        this.jssipUA.sendRequest('SUBSCRIBE', this.target, this.params, headers, body, this, this.credential);
    }
    _getExpires(r) {
        let e = r.getHeader('expires');
        return e ? parseInt(e) : -1;
    }
    _calculateTimeoutMs(expires) {
        return expires >= 140 ? (expires * 1000 / 2) + Math.floor(((expires / 2) - 70) * 1000 * Math.random()) : (expires * 1000) - 5000;
    }
    _scheduleSubscribe(timeout) {
        this.log('CSubs: next SUBSCRIBE will be sent in ' + Math.floor(timeout / 1000) + ' sec');
        this.expiresTimer = setTimeout(() => {
            this.expiresTimer = undefined;
            this._send(null, this.headers);
        }, timeout);
    }
}

/*
    Server SUBSCRIBE dialog

    Not implemented:
    - fetch-SUBSCRIBE (initial SUBSCRIBE with expires:0), see RFC 6665 4.4.3.
 */
class ServerSubscribeDialog {
    constructor({ jssipUA, log, subscribe, contentType, headers, listeners, credential, pending }) {
        if (!jssipUA)
            throw 'missed reference to JsSIP UA instance';
        this.jssipUA = jssipUA;
        this.log = log ? log : console.log;
        this.expiresTS = null;
        this.expiresTimer = null;
        this.state = pending ? 'pending' : 'active';
        this.finalNotifyWasSent = false;
        this.receivedFirstNotifyResponse = false;
        this.id = null;
        this.eventName = subscribe.getHeader('event');
        this.contentType = contentType;
        if (!contentType)
            throw 'missed contentType for NOTIFY'
        this.expires = parseInt(subscribe.getHeader('expires'));
        if (!listeners)
            throw 'missed listeners';
        if (!listeners.subscribe)
            throw 'missed listeners.subscribe()';
        if (!listeners.terminated)
            throw 'missed listeners.terminated()';
        this.listeners = listeners;
        this.credential = credential;
        let user = subscribe.to.uri.user;
        let domain = subscribe.to.uri.host;
        this.contact = '<sip:' + user + '@' + domain + ';transport=ws>';
        this.rcseq = subscribe.cseq;
        this.data = {};
        this.headers = headers ? headers : [];
        this.target = subscribe.from.uri.user;
        subscribe.to_tag = JsSIP.Utils.newTag();
        this.params = {
            from: subscribe.to,
            from_tag: subscribe.to_tag,
            to: subscribe.from,
            to_tag: subscribe.from_tag,
            call_id: subscribe.call_id,
            cseq: Math.floor(Math.random() * 10000 + 1),
        };
        this.id = this.params.call_id + this.params.from_tag + this.params.to_tag;
        this.log('SSubs: add dialog id=' + this.id);
        this.jssipUA.newDialog(this);
        this._setExpiresTS();
        this._setExpiresTimer();
        subscribe.reply(200, null, ['Expires: ' + this.expires, 'Contact: ' + this.contact]);
        this.wasTerminated = false;
        this.terminatedReason = undefined;
    }

    setActiveState() {
        if (this.state === 'pending') {
            this.log('SSubs: state switched from "pending" to "active"');
            this.state = 'active';
        }
    }

    sendNotify(body = null) {
        let subsState = this.state;
        if (this.state !== 'terminated') {
            subsState += ';expires=' + this._getExpiresTS();
        } else if (this.terminatedReason) {
            subsState += ';reason=' + this.terminatedReason;
        }
        let headers = this.headers.slice();
        headers.push('Subscription-State: ' + subsState);
        headers.push('Event: ' + this.eventName);
        if (body) {
            headers.push('Content-Type: ' + this.contentType);
        }
        this.params.cseq++;
        this.jssipUA.sendRequest('NOTIFY', this.target, this.params, headers, body, this, this.credential);
    }

    sendFinalNotify(body = null, reason = null) {
        if (this.finalNotifyWasSent)
            return;
        this.finalNotifyWasSent = true;
        this._dialogTerminated('send final notify');
        this.terminatedReason = reason;
        this.sendNotify(body);
    }

    // NOTIFY callbacks
    onAuthenticated() {
        this.params.cseq++;
    }
    onRequestTimeout() {
        this._dialogTerminated('notify response timeout');
    }
    onTransportError() {
        this._dialogTerminated('notify transport error');
    }
    onReceiveResponse(response) {
        if (response.status_code >= 200 && response.status_code < 300) {
            if (!this.receivedFirstNotifyResponse) {
                this.receivedFirstNotifyResponse = true;
                this.route_set = response.getHeaders('record-route').reverse();
                if (this.route_set.length > 0)
                    this.params.route_set = this.route_set;
            }
        } else if (response.status_code >= 300) {
            this._dialogTerminated('receive notify non-OK response');
        }
    }
    // end of NOTIFY callbacks

    // dialog callback
    receiveRequest(request) {
        if (request.method !== 'SUBSCRIBE') {
            reply(405); // Method Not Allowed    
            return;
        }
        let h = request.getHeader('expires');
        if (h === undefined || h === null) { // SUBSCRIBE without Expires. See RFC 6665 3.1.1
            h = '900'; // Set default expires value
            this.log('SSubs: subscribe without Expires header. Set by default ' + h);
        }
        this.expires = parseInt(h);
        request.reply(200, null, ['Expires: ' + this.expires, 'Contact: ' + this.contact]);

        let body = request.body;
        let ct = request.getHeader('content-type');
        let isUnsubscribe = this.expires === 0;
        this.log('SSubs>>> subscribe: id=' + this.id, body, ct);
        // Upon receive SUBSCRIBE server must send NOTIFY.
        this.listeners.subscribe(this, isUnsubscribe, request, body, ct);

        if (isUnsubscribe) {
            this._dialogTerminated('receive un-subscribe');
        } else {
            this._setExpiresTS();
            this._setExpiresTimer();
        }
    }

    _dialogTerminated(reason) {
        if (this.wasTerminated) // to prevent duplicate call
            return;
        this.wasTerminated = true;
        this.state = 'terminated';
        clearTimeout(this.expiresTimer);
        // If delay needed ?
        setTimeout(() => {
            this.log('SSubs: remove dialog id=' + this.id);
            this.jssipUA.destroyDialog(this);
        }, 32000);
        this.log(`SSubs>>> terminated: "${reason}" id=${this.id}`);
        this.listeners.terminated(this, reason);
    }
    _setExpiresTS() {
        this.expiresTS = new Date().getTime() + this.expires * 1000;
    }
    _getExpiresTS() {
        let current = new Date().getTime();
        let delta = Math.floor((this.expiresTS - current) / 1000);
        return delta >= 0 ? delta : 0;
    }
    _setExpiresTimer() {
        clearTimeout(this.expiresTimer);
        setTimeout(() => {
            if (this.finalNotifyWasSent)
                return;
            this.terminatedReason = 'timeout';
            this.finalNotifyWasSent = true;
            this.sendNotify();
            this._dialogTerminated('subscription expired');
        }, this.expires * 1000);
    }
}
